orttraining/orttraining/python/training/_checkpoint_storage.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# _checkpoint_storage.py
+
+import h5py
+from collections.abc import Mapping
+
+def _dfs_save(group, save_obj):
+    """Recursively go over each level in the save_obj dictionary and save values to a hdf5 group"""
+
+    for key, value in save_obj.items():
+        if isinstance(value, Mapping):
+            subgroup = group.create_group(key)
+            _dfs_save(subgroup, value)
+        else:
+            group[key] = value
+
+def save(save_obj: dict, path):
+    """Saves the input dictionary to a binary file or python file-like object specified by path.
+
+    Saves an hdf5 representation of the save_obj dictionary to a file or a file-like object specified by path.
+
+    Args:
+        save_obj: dictionary that needs to be saved.
+            save_obj should consist of types understandable by hdf5 file format.
+            if hdf5 does not recognize a type, an exception is raised.
+            if save_obj is not a dictionary, a ValueError is raised.
+        path: string representation to a file path or a python file-like object.
+            if file already exists at path, an exception is raised.
+    """
+    if not isinstance(save_obj, Mapping):
+        raise ValueError("Object to be saved must be a dictionary")
+
+    with h5py.File(path, 'w-') as f:
+        _dfs_save(f, save_obj)
+
+def _dfs_load(group, load_obj):
+    """Recursively go over each level in the hdf5 group and load the values into the given dictionary"""
+
+    for key in group:
+        if isinstance(group[key], h5py.Group):
+            load_obj[key] = {}
+            _dfs_load(group[key], load_obj[key])
+        else:
+            load_obj[key] = group[key][()]
+
+def load(path, key=None):
+    """Loads the data stored in the binary file specified at the given path into a dictionary and returns it.
+
+    Loads the data from an hdf5 file specified at the given path into a python dictionary.
+    Loaded dictionary contains numpy equivalents of python data types. For example:
+        tensor -> saved as a numpy array and loaded as a numpy array.
+        bool -> saved as a numpy bool and loaded as a numpy bool
```

Comment:
h5py saving changes he original type. If the original type is a tensor, it is stored as a numpy array, and so on.
In order to avoid changing the types, we can save the user dict in a bytes representation.
Will explicitly add a line that indicates that the types will be changed.